File: frontend/src/app/domain/models/message.ts
export type MessageRole = 'user' | 'assistant';

export interface QueryArtifactColumn {
    readonly name: string;
    readonly type: string;
}

export interface QueryArtifact {
    readonly id: string;
    readonly title: string;
    readonly summary: string;
    readonly type: 'query_result';
    readonly generatedFrom: string;
    readonly sql: string | null;
    readonly columns: QueryArtifactColumn[];
    readonly rows: Record<string, unknown>[];
    readonly rowCount: number;
    readonly truncated: boolean;
    readonly error: string | null;
}

export interface Message {
    readonly id: string;
    readonly role: MessageRole;
    readonly content: string;
    readonly timestamp: Date;
    readonly sql?: string | null;
    readonly error?: string | null;
    readonly artifact?: QueryArtifact | null;
    readonly conversationTitle?: string | null;
}

File: backend/app/domain/entities/message.py
from dataclasses import dataclass
from datetime import datetime
from enum import StrEnum
from typing import Any


class MessageRole(StrEnum):
    USER = "user"
    ASSISTANT = "assistant"


@dataclass(frozen=True, slots=True)
class Message:
    id: str
    role: MessageRole
    content: str
    timestamp: datetime
    sql: str | None = None
    error: str | None = None
    artifact: dict[str, Any] | None = None
    conversation_title: str | None = None
